### PR TITLE
Allow --load with same target, but with different tag names

### PR DESCRIPTION
**Pull request by dchw**
Resolves https://github.com/earthly/earthly/issues/685.

This allows the same upstream target to be used multiple times in a `WITH DOCKER --load`, with different docker tag names. It also oversteps its bounds a little bit and formalizes some of the cache we had set up.

### CODE REVIEW

File: states/solveCache.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+)
+
+// SolveCache is a formal version of the cache we keep mapping targets to their LLB state.
+type SolveCache map[string]llb.State
+
+// StateKeyFunc is a function that can build a key for the SolveCache. These keys seem to be
+// Highly convention based, and used elsewhere too (LocalFolders?), so this is a step at
+// formalizing that convention, since we sometimes need one key, and sometimes another.
+// It may give us some toeholds to help with some refactoring later.
+type StateKeyFunc func() string
```
Comment (dchw) [inline states/solveCache.go:17]: Its late; but the whole idea here was to make the key a type. An alternative here would be to "render" the data into the key, and use this as something like `type StateKey string`. I think thats better actually, how do I request a fix of my own PR? ;)

File: earthfile2llb/withdockerrun.go [hunk 2]
```diff
@@ -341,14 +341,14 @@ func (wdr *withDockerRun) solveImage(ctx context.Context, mts *states.MultiTarge
...
+		keyFunc(),
 		llb.SessionID(sessionID),
 		llb.Platform(llbutil.DefaultPlatform()),
 		llb.WithCustomNamef("[internal] docker tar context %s %s", opName, sessionID),
 	)
 	wdr.tarLoads = append(wdr.tarLoads, tarContext)
-	wdr.c.mts.Final.LocalDirs[solveID] = outDir
-	wdr.c.opt.SolveCache[solveID] = tarContext
+	wdr.c.mts.Final.LocalDirs[keyFunc()] = outDir
+	wdr.c.opt.SolveCache.Set(keyFunc, tarContext)
```
Comment (dchw) [inline earthfile2llb/withdockerrun.go:351]: More or less the problem was what the target hash is not enough here to make assumptions on. We need to also include the name its being assigned.

File: states/solveCache.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+// Delete removes a LLB state from a given solve cache, using the KeyFunc to derive the key
+func (sc *SolveCache) Delete(kf StateKeyFunc, state llb.State) {
+	delete((*sc), kf())
+}
+
+// KeyFromHashAndTag builds a state key from a given target state and a docker tag.
+// This is useful when you want to reference the same image but with a different name.
+func KeyFromHashAndTag(target *SingleTarget, dockerTag string) (StateKeyFunc, error) {
```
Comment (dchw) [inline states/solveCache.go:43]: See aformementioned idea about a special string type.

**Review by dchw — COMMENTED** [review]
I had a better idea on something, but leaving this up for comments. I will make the fix I suggested tomorrow.

**Review by alexcb — APPROVED** [review]

File: states/solveCache.go [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+// KeyFromState is a simple wrapper to get a key from a given state using the hash of its target.
+func KeyFromState(target *SingleTarget) (StateKey, error) {
+	hash, err := target.TargetInput.Hash()
+	if err != nil {
+		return StateKey(""), errors.Wrap(err, "target input hash")
+	}
+
+	return StateKey(hash), nil
+}
```
Comment (vladaionescu) [inline states/solveCache.go:61]: Nit: Name this file `solvecache.go` for consistency?

File: examples/tests/single-target-load-multi-name.earth [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+            --pull ubuntu:latest       # Test that we can load other images easily
+        RUN docker images && \
+            docker run a:latest && \
+            echo "-----" && \
+            docker run b:latest && \
+            echo "-----" && \
+            docker run c:latest uname -a && \
+            echo "-----" && \
+            docker run ubuntu:latest uname-a
+    END
```
Comment (vladaionescu) [inline examples/tests/single-target-load-multi-name.earth:22]: This file is not actually used, right?

**Review by vladaionescu — APPROVED** [review]
Nice find!

File: states/solveCache.go [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+	"github.com/moby/buildkit/client/llb"
+	"github.com/pkg/errors"
+)
+
+// SolveCache is a formal version of the cache we keep mapping targets to their LLB state.
+type SolveCache map[StateKey]llb.State
+
+// StateKey is a type for a key in SolveCache. These keys seem to be highly convention based,
+// and used elsewhere too (LocalFolders?). so this is a step atformalizing that convention,
```
Comment (vladaionescu) [inline states/solveCache.go:14]: Those are actually cache folders rather than pure local folders.

**Review by vladaionescu — APPROVED** [review]

File: examples/tests/single-target-load-multi-name.earth (hunk 5 above)
Comment (dchw) [reply examples/tests/single-target-load-multi-name.earth:22]: No. I thought I removed that when I moved it into the with docker tests. Ill remove it.

File: states/solveCache.go (hunk 4 above)
Comment (dchw) [reply states/solveCache.go:61]: Yep; can do. I <3 consistency

File: states/solveCache.go (hunk 6 above)
Comment (dchw) [reply states/solveCache.go:14]: :+1:

**Pull request merged**